state/unit.go

```diff
@@ -321,6 +321,103 @@ func (u *Unit) PasswordValid(password string) bool {
 	return false
 }
 
+// unitStateDoc records the state persisted by the charm executing in the unit.
+type unitStateDoc struct {
+	// DocID is always the same as a unit's global key.
+	DocID string `bson:"_id"`
+
+	// State encodes the unit's persisted state as a list of key-value pairs.
+	State map[string]string `bson:"state,omitempty"`
+
+	TxnRevno int64 `bson:"txn-revno"`
+}
+
+// removeUnitStateOp returns the operation needed to remove the unit state
+// document associated with the given globalKey.
+func removeUnitStateOp(mb modelBackend, globalKey string) txn.Op {
+	return txn.Op{
+		C:      unitStatesC,
+		Id:     mb.docID(globalKey),
+		Remove: true,
+	}
+}
+
+// SetState persisted the state for a unit.
+func (u *Unit) SetState(unitState map[string]string) error {
+	unitGlobalKey := u.globalKey()
+	txnDoc := struct {
+		TxnRevno int64 `bson:"txn-revno"`
+	}{}
+
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		coll, closer := u.st.db().GetCollection(unitStatesC)
+		defer closer()
+
+		if err := coll.FindId(unitGlobalKey).One(&txnDoc); err != nil {
+			if err != mgo.ErrNotFound {
+				return nil, errors.Trace(err)
+			}
+
+			escapedState := make(map[string]string, len(unitState))
+			for k, v := range unitState {
+				escapedState[mgoutils.EscapeKey(k)] = v
+			}
+
+			return []txn.Op{{
+				C:      unitStatesC,
+				Id:     unitGlobalKey,
+				Assert: txn.DocMissing,
+				Insert: unitStateDoc{
+					DocID: unitGlobalKey,
+					State: escapedState,
+				},
+			}}, nil
+		}
+
+		// State keys may contain dots or dollar chars which need to be escaped.
+		escapedState := make(bson.M, len(unitState))
+		for k, v := range unitState {
+			escapedState[mgoutils.EscapeKey(k)] = v
+		}
```

Comment:
Do we want to protect against the change already being made?  Or no change necessary here?

Reply:
Already made? You mean to diff the existing settings and the new and make the set a no-op?

Reply:
(if nothing needs to change)

Reply:
yes.